File: src/canonical_record.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::audit_chain::{FrozenChain, ReasoningChain};
use crate::trusted_knowledge::TrustedKnowledgeEntry;
use base64::{Engine as _, engine::general_purpose::STANDARD as B64};
use ed25519_dalek::{Signer, SigningKey};
use hex;
use rand_core::OsRng;
use sha2::{Digest, Sha256};

/// A canonical record representation used for Codex Nexus.
///
/// Each record includes a minimal set of metadata fields that uniquely
/// identify and describe the entry regardless of downstream storage
/// technology.  The `payload` holds the domain‑specific data (for
/// example a FrozenChain, a memory block, or a RAG document).  The
/// `hash` is a hex‑encoded SHA256 hash over the canonical plaintext
/// representation of the payload (before encryption).  The `sig` and
/// `pub_key` hold the Ed25519 signature and verifying key for the
/// payload hash.  `witnesses` collects any additional signatures
/// attesting to the record.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CanonicalRecord {
    pub kind: String,
    pub schema_version: u32,
    pub id: String,
    pub tenant: String,
    pub ts: DateTime<Utc>,
    pub space: String,
    pub payload: Value,
    pub links: Vec<Link>,
    pub prev: Option<String>,
    pub hash: String,
    pub sig: Option<String>,
    pub pub_key: Option<String>,
    pub witnesses: Vec<WitnessRecord>,
}

/// Simple relation between two records.  `label` describes the
/// relationship type, `target` is the target record ID.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Link {
    pub label: String,
    pub target: String,
}

/// External witness signature attached to a record.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WitnessRecord {
    pub witness_id: String,
    pub signature: String,
    pub timestamp: DateTime<Utc>,
    pub authority: String,
}

impl CanonicalRecord {
    /// Minimal constructor for tests
    pub fn new_minimal_for_test(id: &str, tenant: &str, space: &str, payload: Value) -> Self {
        Self {
            kind: "test".to_string(),
            schema_version: 1,
            id: id.to_string(),
            tenant: tenant.to_string(),
            ts: chrono::Utc::now(),
            space: space.to_string(),
            payload,
            links: vec![],
            prev: None,
            hash: String::new(),
            sig: None,
            pub_key: None,
            witnesses: vec![],
        }
    }
    /// Construct a canonical record from a FrozenChain.  The caller
    /// provides the tenant and space (e.g., "user" vs "system").  The
    /// `prev` argument links the new record to the previous version
    /// (if any).
    pub fn from_frozen_chain(
        chain: &FrozenChain,
        tenant: &str,
        space: &str,
        prev: Option<&str>,
    ) -> Result<Self, String> {
        // Serialize the FrozenChain as JSON for the payload.  Any
        // serialization error will propagate back to the caller.
        let payload = serde_json::to_value(chain)
            .map_err(|e| format!("Failed to serialize FrozenChain: {e}"))?;

        // Map witnesses from the FrozenChain type to CanonicalRecord.
        let witnesses: Vec<WitnessRecord> = chain
            .witnesses
            .iter()
            .map(|w| WitnessRecord {
                witness_id: w.witness_id.clone(),
                signature: w.signature.clone(),
                timestamp: w.timestamp,
                authority: w.authority.clone(),
            })
            .collect();

        // Build links.  Parent chains are represented as a link of
        // relation type "parent".
        let links: Vec<Link> = chain
            .parent_chain_ids
            .iter()
            .map(|p| Link {
                label: "parent".to_string(),
                target: p.clone(),
            })
            .collect();

        Ok(CanonicalRecord {
            kind: "frozen_chain".to_string(),
            schema_version: 1,
            id: chain.chain_id.clone(),
            tenant: tenant.to_string(),
            ts: chain.frozen_at,
            space: space.to_string(),
            payload,
            links,
            prev: prev.map(String::from),
            hash: chain.content_hash.clone(),
            sig: chain.signature.clone(),
            pub_key: chain.public_key.clone(),
            witnesses,
        })
    }

    /// Construct a canonical record from a ReasoningChain.  
    /// ReasoningChains are mutable process records used during active reasoning.
    pub fn from_reasoning_chain(
        chain: &ReasoningChain,
        tenant: &str,
        space: &str,
        prev: Option<&str>,
    ) -> Result<Self, String> {
        // Serialize the ReasoningChain as JSON for the payload
        let payload = serde_json::to_value(chain)
            .map_err(|e| format!("Failed to serialize ReasoningChain: {e}"))?;

        // ReasoningChains don't have witnesses yet (they're still mutable)
        let witnesses: Vec<WitnessRecord> = Vec::new();

        // ReasoningChains don't have parent_chain_ids (they're still mutable)
        // Links will be established when the chain is frozen
        let links: Vec<Link> = Vec::new();

        // Generate a hash of the canonical payload for integrity
        let canonical_json = serde_json::to_string(&payload)
            .map_err(|e| format!("Failed to serialize payload: {e}"))?;
        let hash = hex::encode(Sha256::digest(canonical_json.as_bytes()));

        Ok(CanonicalRecord {
            kind: "reasoning_chain".to_string(),
            schema_version: 1,
            id: chain.audit.chain_id.clone(),
            tenant: tenant.to_string(),
            ts: Utc::now(), // ReasoningChains are created "now"
            space: space.to_string(),
            payload,
            links,
            prev: prev.map(String::from),
            hash,
            sig: None, // ReasoningChains are not signed yet (still mutable)
            pub_key: None,
            witnesses,
        })
    }

    /// Produce a canonical JSON string for this record using RFC 8785 (JCS).
    /// This ensures cryptographically stable canonicalization for hashing and signing.
    pub fn to_canonical_json(&self) -> Result<String, String> {
        crate::canonicalize::canonicalize_record(self)
    }

    /// Construct a canonical record from a TrustedKnowledgeEntry.  This helper
    /// serializes the entry into a JSON payload, assigns basic metadata
    /// (kind = entry.category, id = entry.id), and signs the canonical
    /// representation using a newly generated Ed25519 key.  The caller
    /// must provide the tenant and space strings to record the origin of
    /// this data.  The schema_version can be chosen by the caller to
    /// enable future migrations.
    pub fn from_trusted_entry(
        entry: &TrustedKnowledgeEntry,
        tenant: &str,
        space: &str,
        schema_version: u32,
    ) -> Result<Self, String> {
        // Convert the TrustedKnowledgeEntry to a JSON value.  Any
        // serialization error is propagated back to the caller.
        let payload = serde_json::to_value(entry)
            .map_err(|e| format!("Failed to serialize TrustedKnowledgeEntry: {e}"))?;

        // Build a provisional record (without hash/signature) so we can
        // canonicalize it for hashing and signing.  Use empty links
        // and witnesses, and None prev for now.
        let mut record = CanonicalRecord {
            kind: entry.category.clone(),
            schema_version,
            id: entry.id.clone(),
            tenant: tenant.to_string(),
            ts: chrono::Utc::now(),
            space: space.to_string(),
            payload,
            links: Vec::new(),
            prev: None,
            hash: String::new(),
            sig: None,
            pub_key: None,
            witnesses: Vec::new(),
        };

        // Canonicalize the JSON for hashing.  This uses the same
        // canonicalization routine as to_canonical_json, sorting keys
        // consistently.  If canonicalization fails, return an error.
        let canon_json = record.to_canonical_json()?;

        // Compute SHA256 hash of the canonical JSON string.
        let mut hasher = Sha256::new();
        hasher.update(canon_json.as_bytes());
        let digest = hasher.finalize();
        let hash_hex = hex::encode(digest);

        // Generate a new signing key for this record.  In a real
        // deployment this key would be loaded from secure storage.
        let signing_key = SigningKey::generate(&mut OsRng);
        let signature = signing_key.sign(canon_json.as_bytes());
        let sig_b64 = B64.encode(&signature.to_bytes());
        let pub_b64 = B64.encode(&signing_key.verifying_key().to_bytes());

        // Populate the hash, signature and public key on the record.
        record.hash = hash_hex;
        record.sig = Some(sig_b64);
        record.pub_key = Some(pub_b64);

        Ok(record)
    }
}

File: cpp/src/canonical_record.cpp
#include "sigil/canonical_record.hpp"
#include "sigil/json_canonicalization.hpp"
#include "sigil/crypto.hpp"
#include "sigil/witness_registry.hpp"
#include <format>
#include <chrono>

namespace sigil
{

    using Json = nlohmann::json;
    using Bytes = crypto::Bytes;

    // ========== Link Implementation ==========

    Json Link::to_json() const
    {
        return Json{
            {"label", label},
            {"target", target}};
    }

    Link Link::from_json(const Json &j)
    {
        return Link{
            j.at("label").get<std::string>(),
            j.at("target").get<std::string>()};
    }

    // ========== WitnessRecord Implementation ==========

    Json WitnessRecord::to_json() const
    {
        return Json{
            {"witness_id", witness_id},
            {"signature", signature},
            {"timestamp", timestamp},
            {"authority", authority}};
    }

    WitnessRecord WitnessRecord::from_json(const Json &j)
    {
        return WitnessRecord{
            j.at("witness_id").get<std::string>(),
            j.at("signature").get<std::string>(),
            j.at("timestamp").get<std::string>(),
            j.at("authority").get<std::string>()};
    }

    // ========== CanonicalRecord Implementation ==========

    CanonicalRecord CanonicalRecord::create(
        const std::string &kind,
        const std::string &id,
        const std::string &tenant,
        const std::string &space,
        const Json &payload)
    {
        CanonicalRecord record;
        record.kind = kind;
        record.schema_version = "1.0";
        record.id = id;
        record.tenant = tenant;
        record.space = space;
        record.payload = payload;

        // Generate ISO 8601 timestamp
        auto now = std::chrono::system_clock::now();
        auto time_t_now = std::chrono::system_clock::to_time_t(now);
        auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(
                      now.time_since_epoch()) %
                  1000;

        std::tm tm_buf;
        gmtime_r(&time_t_now, &tm_buf);
        record.ts = std::format("{:04d}-{:02d}-{:02d}T{:02d}:{:02d}:{:02d}.{:03d}Z",
                                tm_buf.tm_year + 1900,
                                tm_buf.tm_mon + 1,
                                tm_buf.tm_mday,
                                tm_buf.tm_hour,
                                tm_buf.tm_min,
                                tm_buf.tm_sec,
                                static_cast<int>(ms.count()));

        return record;
    }

    Result<CanonicalRecord> CanonicalRecord::new_signed(
        const std::string &kind,
        const std::string &id,
        const std::string &tenant,
        const std::string &space,
        const Json &payload,
        const std::optional<std::string> &prev)
    {
        // Get signing key from KeyManager
        auto key_result = crypto::KeyManager::get_or_create_canon_key();
        if (!key_result)
        {
            return std::unexpected(key_result.error());
        }
        auto key = *key_result;

        // Create record with unsigned data
        CanonicalRecord rec = create(kind, id, tenant, space, payload);

        // Set previous hash if provided
        if (prev)
        {
            rec.previous_hash = *prev;
        }

        // Compute canonical JSON (excludes hash field)
        Json canonical = rec.to_canonical_json();
        // Use RFC 8785 canonicalization (matches Rust canonicalize.rs)
        std::string canonical_str = sigil::json::RFC8785Canonicalizer::canonicalize(canonical);
        Bytes canonical_bytes(canonical_str.begin(), canonical_str.end());

        // Compute hash (SHA256::hash returns SHA256Hash directly, not Result)
        auto hash_array = crypto::SHA256::hash(canonical_bytes);
        rec.hash = crypto::Base64::encode(Bytes(hash_array.begin(), hash_array.end()));

        // Sign canonical bytes (hash is NOT in signature, matches Rust)
        auto [sig, pub_key] = key.sign_canonical_bytes(canonical_bytes);
        rec.signature = sig;
        rec.public_key = pub_key;

        return rec;
    }

    Json CanonicalRecord::to_json() const
    {
        Json j = {
            {"kind", kind},
            {"schema_version", schema_version},
            {"id", id},
            {"tenant", tenant},
            {"ts", ts},
            {"space", space},
            {"payload", payload}};

        // Add links if present
        if (!links.empty())
        {
            Json links_array = Json::array();
            for (const auto &link : links)
            {
                links_array.push_back(link.to_json());
            }
            j["links"] = links_array;
        }

        if (hash)
            j["hash"] = *hash;
        if (signature)
            j["sig"] = *signature; // Match Rust field name
        if (public_key)
            j["pub_key"] = *public_key; // Match Rust field name
        if (previous_hash)
            j["prev"] = *previous_hash; // Match Rust field name

        // Add witnesses if present
        if (!witnesses.empty())
        {
            Json witnesses_array = Json::array();
            for (const auto &witness : witnesses)
            {
                witnesses_array.push_back(witness.to_json());
            }
            j["witnesses"] = witnesses_array;
        }

        return j;
    }

    Result<CanonicalRecord> CanonicalRecord::from_json(const Json &j)
    {
        try
        {
            CanonicalRecord record;

            record.kind = j.at("kind").get<std::string>();
            record.schema_version = j.at("schema_version").get<std::string>();
            record.id = j.at("id").get<std::string>();
            record.tenant = j.at("tenant").get<std::string>();
            record.ts = j.at("ts").get<std::string>();
            record.space = j.at("space").get<std::string>();
            record.payload = j.at("payload");

            if (j.contains("hash"))
            {
                record.hash = j["hash"].get<std::string>();
            }
            if (j.contains("signature") || j.contains("sig"))
            {
                record.signature = j.contains("sig") ? j["sig"].get<std::string>() : j["signature"].get<std::string>();
            }
            if (j.contains("public_key") || j.contains("pub_key"))
            {
                record.public_key = j.contains("pub_key") ? j["pub_key"].get<std::string>() : j["public_key"].get<std::string>();
            }

            // Parse optional fields - support both naming conventions
            if (j.contains("previous_hash"))
            {
                record.previous_hash = j["previous_hash"].get<std::string>();
            }
            else if (j.contains("prev"))
            {
                record.previous_hash = j["prev"].get<std::string>();
            }

            // Parse links
            if (j.contains("links") && j["links"].is_array())
            {
                for (const auto &link_json : j["links"])
                {
                    record.links.push_back(Link::from_json(link_json));
                }
            }

            // Parse witnesses
            if (j.contains("witnesses") && j["witnesses"].is_array())
            {
                for (const auto &witness_json : j["witnesses"])
                {
                    record.witnesses.push_back(WitnessRecord::from_json(witness_json));
                }
            }

            return record;
        }
        catch (const Json::exception &e)
        {
            return std::unexpected(SigilError::invalid_input(
                std::format("Failed to parse CanonicalRecord: {}", e.what())));
        }
    }

    std::string CanonicalRecord::to_canonical_json() const
    {
        // Create JSON without cryptographic fields
        // This matches Rust's canonicalize_record() behavior
        Json j = {
            {"kind", kind},
            {"schema_version", schema_version},
            {"id", id},
            {"tenant", tenant},
            {"ts", ts},
            {"space", space},
            {"payload", payload}};

        // Add links if present
        if (!links.empty())
        {
            Json links_array = Json::array();
            for (const auto &link : links)
            {
                links_array.push_back(link.to_json());
            }
            j["links"] = links_array;
        }

        // Include previous_hash if present (part of the chain)
        // Use "prev" to match Rust
        if (previous_hash)
        {
            j["prev"] = *previous_hash;
        }

        // CRITICAL: Do NOT include hash field (matches Rust map.remove("hash"))
        // Do NOT include sig, pub_key, or witnesses

        // Canonicalize using RFC 8785
        return json::RFC8785Canonicalizer::canonicalize(j);
    }

    std::string CanonicalRecord::compute_hash() const
    {
        auto canonical = to_canonical_json();
        auto hash_bytes = crypto::SHA256::hash(canonical);
        return crypto::SHA256::to_hex(hash_bytes);
    }

    Result<void> CanonicalRecord::sign(const crypto::Ed25519KeyPair &keypair)
    {
        // Compute hash if not already present
        if (!hash)
        {
            hash = compute_hash();
        }

        // Get canonical JSON bytes for signing
        auto canonical = to_canonical_json();
        crypto::Bytes message(canonical.begin(), canonical.end());

        // Sign
        auto sig = keypair.sign(message);

        // Store signature and public key as base64
        signature = crypto::Base64::encode(
            crypto::Bytes(sig.begin(), sig.end()));
        public_key = crypto::Base64::encode(
            crypto::Bytes(keypair.public_key.begin(), keypair.public_key.end()));

        return {};
    }

    bool CanonicalRecord::verify_signature() const
    {
        if (!signature || !public_key)
        {
            return false;
        }

        // Decode signature and public key
        auto sig_result = crypto::Base64::decode(*signature);
        if (!sig_result || sig_result->size() != 64)
        {
            return false;
        }

        auto pubkey_result = crypto::Base64::decode(*public_key);
        if (!pubkey_result || pubkey_result->size() != 32)
        {
            return false;
        }

        // Convert to fixed-size arrays
        crypto::Ed25519Signature sig;
        crypto::Ed25519PublicKey pubkey;
        std::copy(sig_result->begin(), sig_result->end(), sig.begin());
        std::copy(pubkey_result->begin(), pubkey_result->end(), pubkey.begin());

        // Get canonical JSON for verification
        auto canonical = to_canonical_json();
        crypto::Bytes message(canonical.begin(), canonical.end());

        return crypto::Ed25519KeyPair::verify(message, sig, pubkey);
    }

    void CanonicalRecord::add_witness_signature(const WitnessRecord &witness)
    {
        witnesses.push_back(witness);
    }

    void CanonicalRecord::add_witness_signature(
        const std::string &witness_id,
        const std::string &signature,
        const std::string &authority)
    {
        // Generate ISO 8601 timestamp
        auto now = std::chrono::system_clock::now();
        auto time_t_now = std::chrono::system_clock::to_time_t(now);
        auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(
                      now.time_since_epoch()) %
                  1000;

        std::tm tm_buf;
        gmtime_r(&time_t_now, &tm_buf);
        auto timestamp = std::format("{:04d}-{:02d}-{:02d}T{:02d}:{:02d}:{:02d}.{:03d}Z",
                                     tm_buf.tm_year + 1900,
                                     tm_buf.tm_mon + 1,
                                     tm_buf.tm_mday,
                                     tm_buf.tm_hour,
                                     tm_buf.tm_min,
                                     tm_buf.tm_sec,
                                     static_cast<int>(ms.count()));

        witnesses.push_back(WitnessRecord{
            witness_id,
            signature,
            timestamp,
            authority});
    }

    bool CanonicalRecord::verify_witness_signatures() const
    {
        // Without a registry, we cannot validate witness signatures. Enforce failure
        // when witness entries exist to prevent silently trusting unverified signatures.
        return witnesses.empty();
    }

    Result<bool> CanonicalRecord::verify_witness_signatures_with_registry(const WitnessRegistry &registry) const
    {
        if (witnesses.empty())
            return true;

        auto canonical = to_canonical_json();
        crypto::Bytes message(canonical.begin(), canonical.end());

        for (const auto &w : witnesses)
        {
            auto res = registry.validate_witness_signature(w.witness_id, message, w.signature);
            if (!res)
                return std::unexpected(res.error());
            if (!*res)
                return false;
        }
        return true;
    }

} // namespace sigil
